Stripe/STPCustomerContext.swift

```diff
@@ -0,0 +1,372 @@
+            completion(nil, retrieveKeyError)
+          })
+        }
+        return
+      }
+      self.apiClient.retrieveCustomer(using: ephemeralKey) { customer, error in
+        if let customer = customer {
+          customer.updateSources(filteringApplePay: !self.includeApplePayPaymentMethods)
+          self.customer = customer
+        }
+        if let completion = completion {
+          stpDispatchToMainThreadIfNecessary({
+            completion(self.customer, error)
+          })
+        }
+      }
+    })
+  }
+
+  @objc
+  public func updateCustomer(
+    withShippingAddress shipping: STPAddress, completion: STPErrorBlock?
+  ) {
+    keyManager.getOrCreateKey({ ephemeralKey, retrieveKeyError in
+      guard let ephemeralKey = ephemeralKey, retrieveKeyError == nil else {
+        if let completion = completion {
+          stpDispatchToMainThreadIfNecessary({
+            completion(retrieveKeyError)
+          })
+        }
+        return
+      }
+      var params: [String: Any] = [:]
+      params["shipping"] = STPAddress.shippingInfoForCharge(
+        with: shipping,
+        shippingMethod: nil)
+      self.apiClient.updateCustomer(
+        withParameters: params,
+        using: ephemeralKey
+      ) { customer, error in
+        if let customer = customer {
+          customer.updateSources(filteringApplePay: !self.includeApplePayPaymentMethods)
+          self.customer = customer
+        }
+        if let completion = completion {
+          stpDispatchToMainThreadIfNecessary({
+            completion(error)
+          })
+        }
+      }
+    })
+  }
+
+  @objc
+  public func attachPaymentMethod(
+    toCustomer paymentMethod: STPPaymentMethod, completion: STPErrorBlock?
+  ) {
+    keyManager.getOrCreateKey({ ephemeralKey, retrieveKeyError in
+      guard let ephemeralKey = ephemeralKey, retrieveKeyError == nil else {
+        if let completion = completion {
+          stpDispatchToMainThreadIfNecessary({
+            completion(retrieveKeyError)
+          })
+        }
+        return
+      }
+
+      self.apiClient.attachPaymentMethod(
+        paymentMethod.stripeId ?? "",
+        toCustomerUsing: ephemeralKey
+      ) { error in
+        self.clearCachedPaymentMethods()
+        if let completion = completion {
+          stpDispatchToMainThreadIfNecessary({
+            completion(error)
+          })
+        }
+      }
+    })
+  }
+
+  @objc
+  public func detachPaymentMethod(
+    fromCustomer paymentMethod: STPPaymentMethod, completion: STPErrorBlock?
+  ) {
+    keyManager.getOrCreateKey({ ephemeralKey, retrieveKeyError in
+      guard let ephemeralKey = ephemeralKey, retrieveKeyError == nil else {
+        if let completion = completion {
+          stpDispatchToMainThreadIfNecessary({
+            completion(retrieveKeyError)
+          })
+        }
+        return
+      }
+
+      self.apiClient.detachPaymentMethod(
+        paymentMethod.stripeId ?? "",
```

Comment:
mm, doesn't seem like `paymentMethod.stripeId` should be optional

Reply:
++